drivers/crypto/crypto_stm32.c

```diff
@@ -0,0 +1,400 @@
+			crypto_stm32_sessions[i].in_use = true;
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static int crypto_stm32_session_setup(struct device *dev,
+				      struct cipher_ctx *ctx,
+				      enum cipher_algo algo,
+				      enum cipher_mode mode,
+				      enum cipher_op op_type)
+{
+	int ctx_idx;
+	struct crypto_stm32_session *session;
+
+	if (ctx->flags & ~(CRYP_SUPPORT)) {
+		LOG_ERR("Unsupported flag");
+		return -EINVAL;
+	}
+
+	if (algo != CRYPTO_CIPHER_ALGO_AES) {
+		LOG_ERR("Unsupported algo");
+		return -EINVAL;
+	}
+
+	if ((mode != CRYPTO_CIPHER_MODE_ECB) &&
+	    (mode != CRYPTO_CIPHER_MODE_CBC) &&
+	    (mode != CRYPTO_CIPHER_MODE_CTR)) {
+		LOG_ERR("Unsupported mode");
+		return -EINVAL;
+	}
+
+	if ((ctx->keylen != 16U) &&
+	    (ctx->keylen != 24U) &&
+	    (ctx->keylen != 32U)) {
+		LOG_ERR("%u key size is not supported", ctx->keylen);
+		return -EINVAL;
+	}
+
+	ctx_idx = crypto_stm32_get_unused_session_index();
+	if (ctx_idx < 0) {
+		LOG_ERR("No free session for now");
+		return -ENOSPC;
+	}
+	session = &crypto_stm32_sessions[ctx_idx];
+	memset(&session->config, 0, sizeof(session->config));
+
+	switch (ctx->keylen) {
+	case 16U:
+		session->config.KeySize = CRYP_KEYSIZE_128B;
+		break;
+	case 24U:
+		session->config.KeySize = CRYP_KEYSIZE_192B;
+		break;
+	case 32U:
+		session->config.KeySize = CRYP_KEYSIZE_256B;
+		break;
+	}
+
+	if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			session->config.Algorithm = CRYP_AES_ECB;
+			ctx->ops.block_crypt_hndlr = crypto_stm32_ecb_encrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CBC:
+			session->config.Algorithm = CRYP_AES_CBC;
+			ctx->ops.cbc_crypt_hndlr = crypto_stm32_cbc_encrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CTR:
+			session->config.Algorithm = CRYP_AES_CTR;
+			ctx->ops.ctr_crypt_hndlr = crypto_stm32_ctr_encrypt;
+			break;
+		default:
+			break;
+		}
+	} else {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			session->config.Algorithm = CRYP_AES_ECB;
+			ctx->ops.block_crypt_hndlr = crypto_stm32_ecb_decrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CBC:
+			session->config.Algorithm = CRYP_AES_CBC;
+			ctx->ops.cbc_crypt_hndlr = crypto_stm32_cbc_decrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CTR:
+			session->config.Algorithm = CRYP_AES_CTR;
+			ctx->ops.ctr_crypt_hndlr = crypto_stm32_ctr_decrypt;
+			break;
+		default:
+			break;
+		}
+	}
+
+	copy_reverse_words((u8_t *)session->key, CRYPTO_STM32_AES_MAX_KEY_LEN,
+			   ctx->key.bit_stream, ctx->keylen);
+
+	session->config.pKey = crypto_stm32_sessions[ctx_idx].key;
+	session->config.DataType = CRYP_DATATYPE_8B;
+	session->config.DataWidthUnit = CRYP_DATAWIDTHUNIT_BYTE;
+
+	ctx->drv_sessn_state = session;
+	ctx->device = dev;
+
+	return 0;
+}
+
+static int crypto_stm32_session_free(struct device *dev,
+				     struct cipher_ctx *ctx)
+{
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	session->in_use = false;
+
```

Comment:
After free I would expect  HW to be kept under reset, or its clock to be shut down ? 
Here the IP will still be running ...

Reply:
Should we loop over all sessions here and call `HAL_CRYP_DeInit`, if no more session is in use?

Reply:
I think so. 
HAL_CRYP_DeInit() will call HAL_CRYP_MspDeInit(hcryp). 
This will shutdown the IP.
Check that the following is called at the end :   
  __HAL_RCC_CRYP_FORCE_RESET();
  __HAL_RCC_CRYP_RELEASE_RESET();

Reply:
I followed your advice in the latest commit. Thanks.